File: src/samples/Sekiban.Dcb.Orleans.Decider.Wasm.Rs/SekibanDcbDecider.Rust.EventSource/src/events.rs
use sekiban_derive::Event;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct WeatherForecastCreated {
    pub forecast_id: Uuid,
    pub location: String,
    pub date: String,
    pub temperature_c: i32,
    pub summary: String,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct WeatherForecastLocationUpdated {
    pub forecast_id: Uuid,
    pub new_location: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct WeatherForecastDeleted {
    pub forecast_id: Uuid,
    pub deleted_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct StudentCreated {
    pub student_id: Uuid,
    pub name: String,
    pub max_class_count: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct ClassRoomCreated {
    pub class_room_id: Uuid,
    pub name: String,
    pub max_students: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct StudentEnrolledInClassRoom {
    pub student_id: Uuid,
    pub class_room_id: Uuid,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct StudentDroppedFromClassRoom {
    pub student_id: Uuid,
    pub class_room_id: Uuid,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct UserRegistered {
    pub user_id: Uuid,
    pub display_name: String,
    pub email: String,
    pub department: Option<String>,
    pub registered_at: String,
    pub monthly_reservation_limit: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct UserProfileUpdated {
    pub user_id: Uuid,
    pub display_name: String,
    pub email: String,
    pub department: Option<String>,
    pub monthly_reservation_limit: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct UserAccessGranted {
    pub user_id: Uuid,
    pub initial_role: String,
    pub granted_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct UserRoleGranted {
    pub user_id: Uuid,
    pub role: String,
    pub granted_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct RoomCreated {
    pub room_id: Uuid,
    pub name: String,
    pub capacity: i32,
    pub location: String,
    pub equipment: Vec<String>,
    pub requires_approval: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct RoomUpdated {
    pub room_id: Uuid,
    pub name: String,
    pub capacity: i32,
    pub location: String,
    pub equipment: Vec<String>,
    pub requires_approval: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct ReservationDraftCreated {
    pub reservation_id: Uuid,
    pub room_id: Uuid,
    pub organizer_id: Uuid,
    pub organizer_name: String,
    pub start_time: String,
    pub end_time: String,
    pub purpose: String,
    pub selected_equipment: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct ReservationHoldCommitted {
    pub reservation_id: Uuid,
    pub room_id: Uuid,
    pub organizer_id: Uuid,
    pub organizer_name: String,
    pub start_time: String,
    pub end_time: String,
    pub purpose: String,
    pub selected_equipment: Vec<String>,
    pub requires_approval: bool,
    pub approval_request_id: Option<Uuid>,
    pub approval_request_comment: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct ReservationConfirmed {
    pub reservation_id: Uuid,
    pub room_id: Uuid,
    pub organizer_id: Uuid,
    pub organizer_name: String,
    pub start_time: String,
    pub end_time: String,
    pub purpose: String,
    pub selected_equipment: Vec<String>,
    pub confirmed_at: String,
    pub approval_request_id: Option<Uuid>,
    pub approval_request_comment: Option<String>,
    pub approval_decision_comment: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct ReservationCancelled {
    pub reservation_id: Uuid,
    pub room_id: Uuid,
    pub organizer_id: Uuid,
    pub organizer_name: String,
    pub start_time: String,
    pub end_time: String,
    pub purpose: String,
    pub selected_equipment: Vec<String>,
    pub approval_request_comment: Option<String>,
    pub reason: String,
    pub cancelled_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct ReservationRejected {
    pub reservation_id: Uuid,
    pub room_id: Uuid,
    pub organizer_id: Uuid,
    pub organizer_name: String,
    pub start_time: String,
    pub end_time: String,
    pub purpose: String,
    pub selected_equipment: Vec<String>,
    pub approval_request_id: Uuid,
    pub approval_request_comment: Option<String>,
    pub reason: String,
    pub rejected_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct ApprovalFlowStarted {
    pub approval_request_id: Uuid,
    pub reservation_id: Uuid,
    pub room_id: Uuid,
    pub requester_id: Uuid,
    pub approver_ids: Vec<Uuid>,
    pub requested_at: String,
    pub request_comment: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Event)]
#[serde(rename_all = "camelCase")]
pub struct ApprovalDecisionRecorded {
    pub approval_request_id: Uuid,
    pub reservation_id: Uuid,
    pub approver_id: Uuid,
    pub decision: crate::states::ApprovalDecision,
    pub comment: Option<String>,
    pub decided_at: String,
}

File: src/samples/Sekiban.Dcb.Orleans.Decider.Wasm.Go/go-wasm/domain/events.go
package domain

// Weather events
type WeatherForecastCreated struct {
	ForecastId   string `json:"forecastId"`
	Location     string `json:"location"`
	Date         string `json:"date,omitempty"`
	TemperatureC int    `json:"temperatureC"`
	Summary      string `json:"summary"`
	CreatedAt    string `json:"createdAt"`
}

type WeatherForecastLocationUpdated struct {
	ForecastId  string `json:"forecastId"`
	NewLocation string `json:"newLocation"`
	UpdatedAt   string `json:"updatedAt"`
}

type WeatherForecastDeleted struct {
	ForecastId string `json:"forecastId"`
	DeletedAt  string `json:"deletedAt"`
}

// Student/ClassRoom events
type StudentCreated struct {
	StudentId     string `json:"studentId"`
	Name          string `json:"name"`
	MaxClassCount int    `json:"maxClassCount"`
}

type ClassRoomCreated struct {
	ClassRoomId string `json:"classRoomId"`
	Name        string `json:"name"`
	MaxStudents int    `json:"maxStudents"`
}

type StudentEnrolledInClassRoom struct {
	StudentId   string `json:"studentId"`
	ClassRoomId string `json:"classRoomId"`
}

type StudentDroppedFromClassRoom struct {
	StudentId   string `json:"studentId"`
	ClassRoomId string `json:"classRoomId"`
}

// User events
type UserRegistered struct {
	UserId                   string  `json:"userId"`
	DisplayName              string  `json:"displayName"`
	Email                    string  `json:"email"`
	Department               *string `json:"department"`
	RegisteredAt             string  `json:"registeredAt"`
	MonthlyReservationLimit  int     `json:"monthlyReservationLimit"`
}

type UserProfileUpdated struct {
	UserId                   string  `json:"userId"`
	DisplayName              string  `json:"displayName"`
	Email                    string  `json:"email"`
	Department               *string `json:"department"`
	MonthlyReservationLimit  int     `json:"monthlyReservationLimit"`
}

type UserAccessGranted struct {
	UserId      string `json:"userId"`
	InitialRole string `json:"initialRole"`
	GrantedAt   string `json:"grantedAt"`
}

type UserRoleGranted struct {
	UserId    string `json:"userId"`
	Role      string `json:"role"`
	GrantedAt string `json:"grantedAt"`
}

// Room events
type RoomCreated struct {
	RoomId           string   `json:"roomId"`
	Name             string   `json:"name"`
	Capacity         int      `json:"capacity"`
	Location         string   `json:"location"`
	Equipment        []string `json:"equipment"`
	RequiresApproval bool     `json:"requiresApproval"`
}

type RoomUpdated struct {
	RoomId           string   `json:"roomId"`
	Name             string   `json:"name"`
	Capacity         int      `json:"capacity"`
	Location         string   `json:"location"`
	Equipment        []string `json:"equipment"`
	RequiresApproval bool     `json:"requiresApproval"`
}

// Reservation events
type ReservationDraftCreated struct {
	ReservationId     string   `json:"reservationId"`
	RoomId            string   `json:"roomId"`
	OrganizerId       string   `json:"organizerId"`
	OrganizerName     string   `json:"organizerName"`
	StartTime         string   `json:"startTime"`
	EndTime           string   `json:"endTime"`
	Purpose           string   `json:"purpose"`
	SelectedEquipment []string `json:"selectedEquipment"`
}

type ReservationHoldCommitted struct {
	ReservationId          string   `json:"reservationId"`
	RoomId                 string   `json:"roomId"`
	OrganizerId            string   `json:"organizerId"`
	OrganizerName          string   `json:"organizerName"`
	StartTime              string   `json:"startTime"`
	EndTime                string   `json:"endTime"`
	Purpose                string   `json:"purpose"`
	SelectedEquipment      []string `json:"selectedEquipment"`
	RequiresApproval       bool     `json:"requiresApproval"`
	ApprovalRequestId      *string  `json:"approvalRequestId"`
	ApprovalRequestComment *string  `json:"approvalRequestComment"`
}

type ReservationConfirmed struct {
	ReservationId          string   `json:"reservationId"`
	RoomId                 string   `json:"roomId"`
	OrganizerId            string   `json:"organizerId"`
	OrganizerName          string   `json:"organizerName"`
	StartTime              string   `json:"startTime"`
	EndTime                string   `json:"endTime"`
	Purpose                string   `json:"purpose"`
	SelectedEquipment      []string `json:"selectedEquipment"`
	ConfirmedAt            string   `json:"confirmedAt"`
	ApprovalRequestId      *string  `json:"approvalRequestId"`
	ApprovalRequestComment *string  `json:"approvalRequestComment"`
	ApprovalDecisionComment *string `json:"approvalDecisionComment"`
}

type ReservationCancelled struct {
	ReservationId          string   `json:"reservationId"`
	RoomId                 string   `json:"roomId"`
	OrganizerId            string   `json:"organizerId"`
	OrganizerName          string   `json:"organizerName"`
	StartTime              string   `json:"startTime"`
	EndTime                string   `json:"endTime"`
	Purpose                string   `json:"purpose"`
	SelectedEquipment      []string `json:"selectedEquipment"`
	ApprovalRequestComment *string  `json:"approvalRequestComment"`
	Reason                 string   `json:"reason"`
	CancelledAt            string   `json:"cancelledAt"`
}

type ReservationRejected struct {
	ReservationId          string   `json:"reservationId"`
	RoomId                 string   `json:"roomId"`
	OrganizerId            string   `json:"organizerId"`
	OrganizerName          string   `json:"organizerName"`
	StartTime              string   `json:"startTime"`
	EndTime                string   `json:"endTime"`
	Purpose                string   `json:"purpose"`
	SelectedEquipment      []string `json:"selectedEquipment"`
	ApprovalRequestId      string   `json:"approvalRequestId"`
	ApprovalRequestComment *string  `json:"approvalRequestComment"`
	Reason                 string   `json:"reason"`
	RejectedAt             string   `json:"rejectedAt"`
}

// Approval events
type ApprovalFlowStarted struct {
	ApprovalRequestId string   `json:"approvalRequestId"`
	ReservationId     string   `json:"reservationId"`
	RoomId            string   `json:"roomId"`
	RequesterId       string   `json:"requesterId"`
	ApproverIds       []string `json:"approverIds"`
	RequestedAt       string   `json:"requestedAt"`
	RequestComment    *string  `json:"requestComment"`
}

type ApprovalDecisionRecorded struct {
	ApprovalRequestId string  `json:"approvalRequestId"`
	ReservationId     string  `json:"reservationId"`
	ApproverId        string  `json:"approverId"`
	Decision          string  `json:"decision"`
	Comment           *string `json:"comment"`
	DecidedAt         string  `json:"decidedAt"`
}
